lua/zk.lua

```diff
@@ -1,115 +1,77 @@
 
 local M = {}
 
-M.api = require("zk.api")
-
-M.lsp = require("zk.lsp")
-
----The entry point of the plugin
----@param options? table user configuration options
-function M.setup(options)
-  config.options = vim.tbl_deep_extend("force", config.defaults, options or {})
-
-  if config.options.lsp.auto_attach.enabled then
-    util.setup_lsp_auto_attach()
-  end
-
-  if config.options.create_user_commands then
-    vim.cmd([[
-      " Core Commands
-      command! -nargs=? -complete=lua ZkIndex   lua require('zk').index(nil, assert(loadstring('return ' .. <q-args>))())
-      command! -nargs=? -complete=lua ZkNew     lua require('zk').new(nil, assert(loadstring('return ' .. <q-args>))())
-      command! -nargs=? -complete=lua ZkList    lua require('zk').list(nil, assert(loadstring('return ' .. <q-args>))())
-      command! -nargs=? -complete=lua ZkTagList lua require('zk').tag.list(nil, assert(loadstring('return ' .. <q-args>))())
-
-      " Convenience Commands
-      command! -range -nargs=? -complete=lua ZkNewLink lua assert(<range> == 2, "ZkNewLink must be called with '<,'> range. Try :'<'>ZkNewLink"); require('zk').new_link(nil, assert(loadstring('return ' .. <q-args>))())
-    ]])
-    -- The definition of :ZkNewLink is kind of a hack.
-    -- The lua function that is called by ZkNewLink will always use the '<,'> marks to get the selection.
-    -- The only way that this command can be called that makes semantical sense is :'<,'>ZkNewLink.
+local function setup_lsp_auto_attach()
+  --- NOTE: modified version of code in nvim-lspconfig
+  local trigger
+  local filetypes = config.options.lsp.auto_attach.filetypes
+  if filetypes then
+    trigger = "FileType " .. table.concat(filetypes, ",")
+  else
+    trigger = "BufReadPost *"
   end
+  vim.api.nvim_command(string.format("autocmd %s lua require'zk'._lsp_buf_auto_add(0)", trigger))
 end
 
--- Core Commands
-
----Indexes the notebook
+---Automatically called via an |autocmd| if lsp.auto_attach is enabled.
 --
----@param path? string path to explicitly specify the notebook
----@param options table additional options
----@see https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md#zkindex
-function M.index(path, options)
-  M.api.index(path, options, function(stats)
-    vim.notify(vim.inspect(stats))
-  end)
-end
+---@param bufnr number
+function M._lsp_buf_auto_add(bufnr)
+  if vim.api.nvim_buf_get_option(bufnr, "buftype") == "nofile" then
+    return
+  end
 
----Creates a new note
---
----@param path? string path to explicitly specify the notebook
----@param options table additional options
----@see https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md#zknew
-function M.new(path, options)
-  -- neovim does not yet support window/showDocument, therefore we handle options.edit locally
-  if options and options.edit then
-    options.edit = nil -- nil means true in this context
+  if not M.notebook_root(vim.api.nvim_buf_get_name(bufnr)) then
```

Comment:
```suggestion
  if not util.notebook_root(vim.api.nvim_buf_get_name(bufnr)) then
```
paired with the following at the top of the file:
```lua
local util = require("zk.util")
```
It does not seem to introduce any kind of circular dependency.

I think that’s the cause of the slowdown I mentioned earlier. The autocommand was entering some kind of loop, due to the call to the `nil` field (`M.notebook_root`).

Reply:
thanks, this is fixed now